File: DataStructures/Python/Queue.py
class Node:
    def __init__(self, val) -> None:
        self.node = val
        self.next = None

class Queue:
    def __init__(self):
        self.head = None
        self.tail = None

    def append(self, val):
        new_node = Node(val)
        new_node.next = None

        if self.head is None:
            self.head = new_node
            self.tail = new_node
            return

        elif self.head == self.tail:
            self.tail = new_node
            self.head.next = self.tail
        
        else:
            self.tail.next = new_node
            self.tail = new_node

    
    def pop(self):
        if self.head is None:
            return
        self.head = self.head.next

    def __str__(self) -> str:
        result = ""
        temp = self.head
        while(temp != None):
            if temp.next == None:
                result += f"{temp.node}"
            else:
                result += f"{temp.node} ->"
            temp = temp.next
    

        return result


if __name__ == "__main__":
    a = Queue()
    a.append(1)
    print(a)
    a.append(2)
    print(a)
    a.append(3)
    print(a)
    a.append(5)
    print(a)
    a.pop()
    print(a)
    a.append(7)
    print(a)

    from collections import deque
    b = deque()
    b.append(1)
    print(b)
    b.append(2)
    b.append(3)
    b.append(5)
    print(b)
    b.popleft() # This behaviour would signal queue
    print(b)
    b.popleft() # This behaviour would signal queue
    print(b)

File: DataStructures/Java/Queue.java
package Java;
//If they are in the same package java will recognize the object and the next import is not required, if not use the package declaration in
// the project to specify where the object originates from.
// import Java.GenericNode;

public class Queue<T>{
    public GenericNode<T> head;
    public GenericNode<T> tail;
    public Queue(){
        head = null;
        tail = null;
    }

    public void append(T val){
        GenericNode<T> new_node = new GenericNode<>(val);
        if (head == null){
            head = new_node;
            tail = head;
        }
        else if (head == tail){
            tail = new_node;
            head.next = tail;
        }
        else{
            tail.next = new_node;
            tail = new_node;
        }
    }

    public T pop(){
        if(head == null)
            return null;
        
        T temp = head.head;
        if (head == tail){
            head = null;
            tail = null;
        }
        else{
            head = head.next;
        }

        return temp;
    }

    public void print(){
        if(head == null){
            System.out.println("Stack is Empty");
            return;
        }
        GenericNode<T> temp = head;
        while (temp != null){
            if (temp.next == null)
                System.out.print(temp.head);
            else
                System.out.print(temp.head + " -> ");
            temp = temp.next;
        }
        System.out.println();

    }

    public static void main(String[] args) {
        Queue<Integer> myList = new Queue<>();

        myList.append(1);
        myList.append(2);
        myList.append(3);
        myList.print();
        myList.pop();
        myList.print();
        myList.pop();
        myList.pop();
        myList.pop();
        myList.print();
        myList.append(5);
        myList.print();
    }
}